SDL_net.h

```diff
@@ -68,1014 +62,1208 @@ extern "C" {
 
 /**
- * Add a socket to a socket set, to be checked for available data.
+ * Dispose of a previously-created stream socket.
  *
- * Generally you don't want to call this generic function, but rather the
- * specific, inline function that wraps it: SDLNet_TCP_AddSocket() or
- * SDLNet_UDP_AddSocket().
+ * This will immediately disconnect the other side of the connection, if
+ * necessary. Further attempts to read or write the socket on the remote
+ * end will fail.
  *
- * This function will fail if you add a socket to the set when the set already
- * has its maximum number of sockets added, but otherwise it will always
- * succeed.
+ * This will _abandon_ any data queued for sending that hasn't made it
+ * to the socket. If you need this data to arrive, you should wait for it
+ * to transmit before destroying the socket with
+ * SDLNet_GetStreamSocketPendingWrites() or
+ * SDLNet_WaitUntilStreamSocketDrained(). Any data that has arrived from the
+ * remote end of the connection that hasn't been read yet is lost.
  *
- * If `sock` is NULL, nothing is added to the set; this lets you query the
- * number of sockets currently contained in the set.
+ * \param sock stream socket to destroy
  *
- * \param set the socket set to add a new socket to.
- * \param sock the socket to add to the set.
- * \returns the total number of sockets contained in the set (including this
- *          new one), or -1 if the set is already full.
+ * \threadsafety You should not operate on the same socket from multiple
+ *               threads at the same time without supplying a serialization
+ *               mechanism. However, different threads may access different
+ *               sockets at the same time without problems.
  *
- * \since This function is available since SDL_net 2.0.0.
+ * \since This function is available since SDL_Net 3.0.0.
  *
- * \sa SDLNet_TCP_AddSocket
- * \sa SDLNet_UDP_AddSocket
- * \sa SDLNet_DelSocket
- * \sa SDLNet_TCP_DelSocket
- * \sa SDLNet_UDP_DelSocket
- * \sa SDLNet_CheckSockets
+ * \sa SDLNet_CreateClient
+ * \sa SDLNet_AcceptClient
+ * \sa SDLNet_GetStreamSocketPendingWrites
+ * \sa SDLNet_WaitUntilStreamSocketDrained
  */
-extern DECLSPEC int SDLCALL SDLNet_AddSocket(SDLNet_SocketSet set, SDLNet_GenericSocket sock);
+extern DECLSPEC void SDLCALL SDLNet_DestroyStreamSocket(SDLNet_StreamSocket *sock);  /* Destroy your sockets when finished with them. Does not block, handles shutdown internally. */
+
+
+/* Datagram (UDP) API... */
+
+typedef struct SDLNet_DatagramSocket SDLNet_DatagramSocket;  /**< a UDP socket. Unreliable, packet-based transmission, with the usual pros/cons */
+
+typedef struct SDLNet_Datagram
+{
+    SDLNet_Address *addr;  /**< this is unref'd by SDLNet_DestroyDatagram. You only need to ref it if you want to keep it. */
+    Uint16 port;  /**< these do not have to come from the same port the receiver is bound to. */
+    Uint8 *buf;
+    int buflen;
```

Comment:
Building with `-Wall`, I get a few warnings about signed/unsigned comparisons.
e.g. this `buflen`, should it be `size_t` instead, or an unsigned SDL `UintXX` type?
If you agree, some arguments that accept a buffer lenght might also need a similar change.

Reply:
I have real questions about signed vs unsigned and size_t vs int in our APIs...

Reply:
There was some thinking around the end of 2022, when we first started working on SDL3, that we shouldn't used unsigned for things that can reasonably avoid it, because you still have can have an underflow, but a negative 1 is much easier to flag as an invalid parameter than a 0xFFFFFFFF.

There was also a discussion somewhere about how size_t was actually bad, too.

I'm happy to change this to whatever (and adjust SDL3's style to match, too).

Reply:
I just pushed a commit that updates the cmake script and adds a VisualC solution (untested).
SDL_net is currently built without `-Werror` (`-DSDL3NET_WERROR=OFF`) on ci, so you can have a look at the warnings.
It's probably just a matter of adding a few casts.


https://github.com/madebr/SDL_net/actions/runs/6318667797

Reply:
> I just pushed a commit that updates the cmake script

Can you also request c99 through cmake? No other sdl3 versions of satellite libs does that but this one is actively using c99ism now.

Reply:
> There was also a discussion somewhere about how size_t was actually bad, too.

Well, it's unsigned, so it has the drawbacks of unsigned..

I don't think size_t would make much sense here, a UDP datagram usually holds <1500 bytes, maybe up to 10000 bytes or so with jumbo frames, so a normal (32bit signed) int (that goes up to 2GB) is more than enough, probably even int16 would suffice (though who knows if the future brings packet sizes >32kB)

Reply:
Sure.
All example, except `get-local-addrs`, need c99 too.
Let's keep `get-local-addrs` c90 to test compatibility.

Reply:
@sezero 
I force pushed this change.
(Sorry if I messed up your workflow ryan)

Reply:
> @sezero I force pushed this change. 

Doesn't work. I believe you intended `${sdl3_net_target_name}` and not `${example}` in there:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e497d11..7dbac5d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -80,6 +80,5 @@ endif()
 add_library(${sdl3_net_target_name} src/SDL_net.c)
 add_library(SDL3_net::${sdl3_net_target_name} ALIAS ${sdl3_net_target_name})
-set_property(TARGET ${example} PROPERTY C_STANDARD 99)
-set_property(TARGET ${example} PROPERTY C_EXTENSIONS FALSE)
+set_property(TARGET ${sdl3_net_target_name} PROPERTY C_STANDARD 99)
 if(NOT TARGET SDL3_net::SDL3_net)
     add_library(SDL3_net::SDL3_net ALIAS ${sdl3_net_target_name})
```

With the above patch applied it builds, with the following warnings:
```
src/SDL_net.c: In function ‘SDLNet_Quit’:
src/SDL_net.c:405: warning: comparison between signed and unsigned integer expressions
src/SDL_net.c: In function ‘SDLNet_ResolveHostname’:
src/SDL_net.c:467: warning: comparison between signed and unsigned integer expressions
src/SDL_net.c: In function ‘SDLNet_WriteToStreamSocket’:
src/SDL_net.c:1098: warning: comparison between signed and unsigned integer expressions
src/SDL_net.c: In function ‘SDLNet_SendDatagram’:
src/SDL_net.c:1366: warning: comparison between signed and unsigned integer expressions
src/SDL_net.c: In function ‘SDLNet_DestroyDatagramSocket’:
src/SDL_net.c:1515: warning: comparison between signed and unsigned integer expressions
src/SDL_net.c: In function ‘SDLNet_WaitUntilInputAvailable’:
src/SDL_net.c:1549: warning: comparison between signed and unsigned integer expressions
examples/voipchat.c: In function ‘SendClientAudioToServer’:
examples/voipchat.c:93: warning: dereferencing type-punned pointer will break strict-aliasing rules
examples/voipchat.c: In function ‘mainloop’:
examples/voipchat.c:140: warning: comparison between signed and unsigned integer expressions
examples/voipchat.c:152: warning: comparison between signed and unsigned integer expressions
examples/voipchat.c:227: warning: dereferencing type-punned pointer will break strict-aliasing rules
examples/voipchat.c: In function ‘main’:
examples/voipchat.c:362: warning: comparison between signed and unsigned integer expressions
```

Note: disabling extensions is harmful, at least on my glibc-2.12 setup.

Reply:
Oops, thanks for seeing the copy/paste error. I got rid of all `C_EXTENSIONS`.

Reply:
>  I got rid of all `C_EXTENSIONS`.

FWIW, the programs seem to build fine w/o extensions. The library doesn't, because some of the types and constants are guarded by `__USE_GNU` in glibc headers.

Reply:
Re-added them. These are also used in the tests of SDL3 and are not causing issues there.

Reply:
> (Sorry if I messed up your workflow ryan)

You did not.  :)

Reply:
> I don't think size_t would make much sense here

I agree, but maybe the into should be Sint32.

UDP has an absolute maximum of 64k per packet, but in practice no one does this, because the theory is bad routers will fragment the packet and risk dropping it by dropping any fragment. So we're all still using the maximum that Quake 3 used in the 1990s of 1500 bytes. Afaik, this is still considered the correct number, a quarter of a century later.

But yeah, a 32-bit int will cover it easily.

Reply:
FYI, the maximum safe MTU is 1272 (the Netgear N450 router drops packets larger than this) and the typical maximum MTU is 1472 (ethernet frame size of 1500 - IP and UDP header size)

Reply:
Ah, yes, I was thinking of the total frame size, not the payload.

Reply:
>  the maximum safe MTU is 1272 (the Netgear N450 router drops packets larger than this)

WTF, that's really broken.. and now imagine using a VPN (that adds its own overhead with own headers) through that thing..

> So are we good with this if I change the buffer sizes to Sint32 instead of int? Or should we just leave it?

I think just int should be fine - I don't think there's any platform relevant for SDL where int is <32bit, and generally it should be a type that's handled well (without much conversion) by the target CPU

Reply:
> the typical maximum MTU is 1472 (ethernet frame size of 1500 - IP and UDP header size)

At least for IPv4, IP+UDP header are 20 bytes (12+8) - with IPv6, it's 48 bytes (40+8).
However, for DSL internet connections there's an additional overhead of 8 bytes for the PPPoe header => 28 bytes for IPv4 => MTU of 1472 (56 bytes => 1444 for IPv6)

And, while at it, OpenVPN has an additional overhead of [69 bytes per packet](https://openvpn-users.narkive.com/jzEE4wYX/overhead-added-to-each-packet-by-openvpn) (+28 = 97 for IPv6), WireGuard has an overhead of [80 bytes per packet](https://lists.zx2c4.com/pipermail/wireguard/2017-December/002201.html), no idea about other VPN solutions, but they should be in the same ballpark.
So when sending UDP through a WireGuard VPN, and using IPv6 for both the connection between VPN client and server, and the tunneled connection, and all that using a DSL-connection, the overhead is: **8 bytes** (DSL/PPPoe) **+ 80 bytes** (WireGuard incl "outer" IPv6 and UDP headers) **+ 48 bytes** ("inner" IPv6 + UDP) **= 136 bytes**, leaving about **1364 bytes** for your data (17 bytes less when using OpenVPN instead of WireGuard).
When using the shitty netgear router @slouken mentioned, it would probably be **1272 bytes - 40 bytes** (WireGuard, but without the outer IP+UDP headers, assuming they're included in the 1272 bytes MTU) **- 48 bytes** (inner IPv6 + UDP) **= 1184 bytes** for your data (or 1167 with OpenVPN)..